hal/device.c

```diff
@@ -0,0 +1,19 @@
+#include "device.h"
+#include <stdio.h>
+#include <stdlib.h>
+
+// TODO: binding device using hal not compiler
+device_t *hardware_deps_init(void) {
```

Comment:
It looks ugly. Can you use something like Linux VFS for registering machine-dependent entries?